### ISSUE TITLE
[Support]: recipe to update built-in seq without warnings?

### ISSUE THREAD

**Issue by shassard** [progfolio/elpaca#216]
### Confirmation

- [X] I have read the documentation and need clarification.
- [X] I have searched previous issues to see if my question is a duplicate.
- [ ] I am  checking these without reading them.
- [X] I have checked the Elpaca github wiki to see if my issue is covered.

### Elpaca Version

```
Elpaca 67e4175 HEAD -> master, origin/master, origin/HEAD
installer:      0.6
emacs-version:  GNU Emacs 29.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)
git --version:  git version 2.43.0
```

### Operating System

arch linux

### Description

I'm trying to use elpaca to upgrade the built-in `seq` package that's required by `magit` (via the `transient` package), as per:
https://github.com/magit/transient/blob/main/lisp/transient.el#L50

The recipe that I'm using is:
```
(elpaca '(seq :type git :host nil :repo "https://git.savannah.gnu.org/git/emacs/elpa.git" :branch "externals/seq")
  (progn (unload-feature 'seq t) (require 'seq)))
```

Which works well enough, other than a warning that's thrown on startup:
```
⛔ Warning (emacs): seq loaded before Elpaca activation
```

Is it possible to suppress that warning, or do you recommend any other way to update built-ins with elpaca?

**Comment by progfolio** [comment]
Thanks for taking the time to fill out a support ticket.

> Is it possible to suppress that warning, or do you recommend any other way to update built-ins with elpaca?

The `BODY` of your declaration is executed after the package is activated.
That's why the warning is still signaled despite unloading seq.
For now you should be able to add a custom build step which unloads seq before it's activated.
See the following test case:

<details open><summary>Test Case</summary>

[How to run this test?](https://github.com/progfolio/elpaca/wiki/Troubleshooting#the-elpaca-test-macro)

```emacs-lisp
(elpaca-test
  ;; Don't mind this setting. Just here to make the test less wasteful.
  :early-init (setq elpaca-menu-functions '(elpaca-menu-gnu-devel-elpa))
  :init
  ;; Define a build step which unloads seq.
  ;; I will likely make `elpaca--continue-build' a public function soon.
  (defun +elpaca-unload-seq (e) (unload-feature 'seq t) (elpaca--continue-build e))
  ;; This will work so long as `elpaca--activate-package' remains the last build step.
  ;; This is another area I'd like to improve by possibly adding better ways to insert/remove build steps. 
  ;; For now, we add the custom build step just before `elpaca--activate-package'. 
  (elpaca `(seq :build ,(append (butlast elpaca-build-steps)
                                (list '+elpaca-unload-seq 'elpaca--activate-package))))
  ;; The rest is only necessary for the test as well...
  (elpaca-wait)
  (elpaca-test-log "seq |"))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>67e4175 HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.6</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0) of 2023-12-18</td>
<tr><td>git</td><td>git version 2.43.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
  INFO     Scraping files for loaddefs... 
  INFO     Scraping files for loaddefs...done
  GEN      ../elpaca-autoloads.el
Cloning into '/tmp/elpaca.fju1ST/elpaca/repos/elpaca'...
Your branch is up to date with 'origin/master'.
Checking /tmp/elpaca.fju1ST/elpaca/repos/elpaca...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-info.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-log.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-manager.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-menu-elpa.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-menu-melpa.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-menu-org.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-process.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-test.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca-ui.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/elpaca.el...
Checking /tmp/elpaca.fju1ST/elpaca/repos/elpaca/doc...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/doc/early-init.el...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/doc/init.el...
Checking /tmp/elpaca.fju1ST/elpaca/repos/elpaca/extensions...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/extensions/elpaca-use-package.el...
Checking /tmp/elpaca.fju1ST/elpaca/repos/elpaca/images...
Checking /tmp/elpaca.fju1ST/elpaca/repos/elpaca/test...
Compiling /tmp/elpaca.fju1ST/elpaca/repos/elpaca/test/elpaca-test.el...
Done (Total of 11 files compiled, 3 skipped in 4 directories)

Downloading GNU-devel ELPA...
elpaca-log: "seq |"
"
seq                            cloning              Cloning                                                                          00.634761
seq                            cloning              Cloning into '/tmp/elpaca.fju1ST/elpaca/repos/seq'...                            00.636052
seq                            cloning              Resolving deltas: 100% (1/1), done.                                              01.014013
seq                            checking-out-ref     Checking out externals/seq                                                       01.017113
seq                            checking-out-ref     Reset branch 'externals/seq'                                                     01.019383
seq                            checking-out-ref     branch 'externals/seq' set up to track 'origin/externals/seq'.                   01.019975
seq                            checking-out-ref     Your branch is up to date with 'origin/externals/seq'.                           01.020472
seq                            ref-checked-out      externals/seq checked out                                                        01.020999
seq                            unblocked            No external dependencies                                                         01.021467
seq                            unblocked            Checking dependency versions                                                     01.021805
seq                            linking              Linking build files                                                              01.022140
seq                            linking              Build files linked                                                               01.022741
seq                            autoloads            Generating autoloads: /tmp/elpaca.fju1ST/elpaca/builds/seq                       01.023303
seq                            autoloads              INFO     Scraping files for loaddefs...                                        01.126492
seq                            autoloads              INFO     Scraping files for loaddefs...done                                    01.129372
seq                            autoloads              GEN      seq-autoloads.el                                                      01.133093
seq                            autoloads            Autoloads Generated                                                              01.133783
seq                            byte-compilation     Byte compiling                                                                   01.134383
seq                            byte-compilation     Checking /tmp/elpaca.fju1ST/elpaca/builds/seq...                                 01.201053
seq                            byte-compilation     Compiling /tmp/elpaca.fju1ST/elpaca/builds/seq/seq-24.el...                      01.211860
seq                            byte-compilation     In seq-drop:                                                                     01.212447
seq                            byte-compilation     seq-24.el:76:8: Warning: function seq-drop used to take 1 or more arguments, now takes 2 01.212953
seq                            byte-compilation     In seq-take:                                                                     01.214592
seq                            byte-compilation     seq-24.el:88:8: Warning: function seq-take used to take 1 or more arguments, now takes 2 01.215092
seq                            byte-compilation     In seq-drop-while:                                                               01.215584
seq                            byte-compilation     seq-24.el:98:8: Warning: function seq-drop-while used to take 2 or more arguments, now takes 2 01.216065
seq                            byte-compilation     seq-24.el:98:2: Warning: docstring wider than 80 characters                      01.216567
seq                            byte-compilation     In seq-take-while:                                                               01.217046
seq                            byte-compilation     seq-24.el:105:2: Warning: docstring wider than 80 characters                     01.217596
seq                            byte-compilation     In seq-filter:                                                                   01.218102
seq                            byte-compilation     seq-24.el:112:2: Warning: docstring wider than 80 characters                     01.218577
seq                            byte-compilation     In seq-remove:                                                                   01.219070
seq                            byte-compilation     seq-24.el:132:2: Warning: docstring wider than 80 characters                     01.219570
seq                            byte-compilation     In seq-some:                                                                     01.220074
seq                            byte-compilation     seq-24.el:153:2: Warning: docstring wider than 80 characters                     01.220575
seq                            byte-compilation     In seq-every-p:                                                                  01.221070
seq                            byte-compilation     seq-24.el:175:2: Warning: docstring wider than 80 characters                     01.221408
seq                            byte-compilation     In seq-count:                                                                    01.221729
seq                            byte-compilation     seq-24.el:183:2: Warning: docstring wider than 80 characters                     01.222037
seq                            byte-compilation     In seq-empty-p:                                                                  01.222341
seq                            byte-compilation     seq-24.el:191:8: Warning: function seq-empty-p used to take 1 or more arguments, now takes 1 01.222646
seq                            byte-compilation     In seq-sort:                                                                     01.222966
seq                            byte-compilation     seq-24.el:197:8: Warning: function seq-sort used to take 2 or more arguments, now takes 2 01.223279
seq                            byte-compilation     In seq-set-equal-p:                                                              01.223605
seq                            byte-compilation     seq-24.el:222:2: Warning: docstring wider than 80 characters                     01.223938
seq                            byte-compilation     seq-24.el:225:38: Warning: ‘seq-contains’ is an obsolete function (as of 27.1); use ‘seq-contains-p’ instead. 01.224259
seq                            byte-compilation     seq-24.el:226:38: Warning: ‘seq-contains’ is an obsolete function (as of 27.1); use ‘seq-contains-p’ instead. 01.224606
seq                            byte-compilation     In seq-uniq:                                                                     01.225000
seq                            byte-compilation     seq-24.el:239:8: Warning: function seq-uniq used to take 1 or more arguments, now takes 1 or 2 01.225357
seq                            byte-compilation     seq-24.el:244:16: Warning: ‘seq-contains’ is an obsolete function (as of 27.1); use ‘seq-contains-p’ instead. 01.225697
seq                            byte-compilation     In seq-partition:                                                                01.236308
seq                            byte-compilation     seq-24.el:303:2: Warning: docstring wider than 80 characters                     01.236785
seq                            byte-compilation     In seq-intersection:                                                             01.237143
seq                            byte-compilation     seq-24.el:318:22: Warning: ‘seq-contains’ is an obsolete function (as of 27.1); use ‘seq-contains-p’ instead. 01.237518
seq                            byte-compilation     In seq-difference:                                                               01.237883
seq                            byte-compilation     seq-24.el:328:27: Warning: ‘seq-contains’ is an obsolete function (as of 27.1); use ‘seq-contains-p’ instead. 01.248531
seq                            byte-compilation     In seq--drop-while-list:                                                         01.248993
seq                            byte-compilation     seq-24.el:402:2: Warning: docstring wider than 80 characters                     01.249402
seq                            byte-compilation     In seq--count-successive:                                                        01.249791
seq                            byte-compilation     seq-24.el:417:2: Warning: docstring wider than 80 characters                     01.250145
seq                            byte-compilation     In seq--make-bindings:                                                           01.250494
seq                            byte-compilation     seq-24.el:428:2: Warning: docstring wider than 80 characters                     01.250890
seq                            byte-compilation     Compiling /tmp/elpaca.fju1ST/elpaca/builds/seq/seq-25.el...                      01.251252
seq                            byte-compilation     Compiling /tmp/elpaca.fju1ST/elpaca/builds/seq/seq-autoloads.el...               01.255473
seq                            byte-compilation     Compiling /tmp/elpaca.fju1ST/elpaca/builds/seq/seq.el...                         01.255945
seq                            byte-compilation     Done (Total of 3 files compiled, 1 skipped)                                      01.256299
seq                            byte-compilation     Byte compilation complete                                                        01.256660
seq                            info                 Compiling Info files                                                             01.256999
seq                            info                 Info source files not found                                                      01.257392
seq                            activation           Activating package                                                               01.260703
seq                            activation           Package build dir added to load-path                                             01.261136
seq                            activation           Caching autoloads                                                                01.261517
seq                            activation           Autoloads cached                                                                 01.261965
seq                            finished             ✓ 0.641 secs                                                                     01.262327
"

 Test Env

Elpaca 67e4175 HEAD -> master, origin/master, origin/HEAD
installer:      0.6
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0)
 of 2023-12-18
git --version:  git version 2.43.0
```
</details>

Does that help?

**Comment by shassard** [comment]
The just about seems to work, but fails on restarting emacs. Though I'm not sure why the `seq` repo is trying to re-clone on every startup.

```
log:
  [2023-12-23 20:35:41] Package queued
  [2023-12-23 20:35:42] Continued by: elpaca--continue-dependency
  [2023-12-23 20:35:42] Cloning
  [2023-12-23 20:35:42] $git clone --depth 1 --single-branch --branch externals/seq https://git.savannah.gnu.org/git/emacs/elpa.git /home/stephen/.config/emacs/elpaca/repos/seq/
  [2023-12-23 20:35:42] fatal: destination path '/home/stephen/.config/emacs/elpaca/repos/seq' already exists and is not an empty directory.
  [2023-12-23 20:35:42] Re-cloning with recipe :depth nil
  [2023-12-23 20:35:42] Continued by: elpaca--clone-process-sentinel
  [2023-12-23 20:35:42] Cloning
  [2023-12-23 20:35:42] $git clone --single-branch --branch externals/seq https://git.savannah.gnu.org/git/emacs/elpa.git /home/stephen/.config/emacs/elpaca/repos/seq/
  [2023-12-23 20:35:42] fatal: destination path '/home/stephen/.config/emacs/elpaca/repos/seq' already exists and is not an empty directory.
  [2023-12-23 20:35:42] fatal: destination path '/home/stephen/.config/emacs/elpaca/repos/seq' already exists and is not an empty directory.
```

**Comment by progfolio** [comment]
Ah, that's a mistake on my part.
Explicitly setting the `:build` to those steps in the recipe will cause them to be run every time.
We would want to conditionally set them.
I'm just about to bed, but I'll come up with something better tomorrow.

**Comment by progfolio** [comment]
This should work better:

```emacs-lisp
  (defun +elpaca-unload-seq (e)
    (and (featurep 'seq) (unload-feature 'seq t))
    (elpaca--continue-build e))

  (elpaca `(seq :build (append (butlast (if (file-exists-p (expand-file-name "seq" elpaca-builds-directory))
                                            elpaca--pre-built-steps
                                          elpaca-build-steps))
                               (list '+elpaca-unload-seq 'elpaca--activate-package)))))
```

**Comment by shassard** [comment]
That worked really well. Thanks!

Here's what I ended up with the final snippet in my `init.el`:
```
(defun +elpaca-unload-seq (e) "Unload seq before continuing the elpaca build, then continue to build the recipe E."
  (and (featurep 'seq) (unload-feature 'seq t))
  (elpaca--continue-build e))
(elpaca `(seq :build ,(append (butlast (if (file-exists-p (expand-file-name "seq" elpaca-builds-directory))
                                          elpaca--pre-built-steps
                                        elpaca-build-steps))
                             (list '+elpaca-unload-seq 'elpaca--activate-package))))
```

**Comment by hankertrix** [comment]
Is it possible to use the use-package macro for this?

```elisp
(elpaca `(seq :build ,(+elpaca-seq-build-steps)))
```

**Comment by progfolio** [comment]
> Is it possible to use the `use-package` macro for this?
> 
> ```elisp
> (elpaca `(seq :build ,(+elpaca-seq-build-steps)))
> ```


Yes. Assuming `elpaca-use-package-mode` is enabled, the following should work:

``` emacs-lisp
(use-package seq :elpaca `(seq :build ,(+elpaca-seq-build-steps)))
```

**Comment by hankertrix** [comment]
That works perfectly! Thank you!

**Comment by ashlineldridge** [comment]
@progfolio Will this be the recommended way of upgrading built-in packages going forwards? While I understand that Elpaca is now doing the _correct_ thing w.r.t. versioned dependencies, it feels like this warrants a more ergonomic/comprehensive solution. Do you see Elpaca eventually abstracting over the build steps [proposed above](https://github.com/progfolio/elpaca/issues/216#issuecomment-1868444883) - e.g. by exposing a new keyword? It doesn't feel quite right to have to reference a bunch of internal/private Elpaca functions to achieve this.

Also, I am not currently able to apply the solution proposed for seq to other built-in packages such as Eldoc (recent versions of Eglot require a higher version of Eldoc than is provided in Emacs 29.1). Elpaca clones the entirety of https://github.com/emacs-mirror/emacs into the local `elpaca/repos/eldoc` directory but something happens with the autoloads so that dependent code can't locate any of the Eldoc functions.

Is there a way to revert to the old behaviour and ignore version requirements for a particular package? As an example of where this would be useful, the `pass` package [requires version 0.1.5 of the password-store-otp package](https://github.com/NicolasPetton/pass/blob/ed7031c5c33a384d07da2d15c9d5f854027a26a2/pass.el#L9) but AFAICT no such version exists (latest version is 0.1.0 and the code hasn't been updated for 3 years). Previously the `pass` package worked fine for me despite it's seemingly unnecessary version constraint but with the recent changes to Elpaca I'm unable to install it.

**Comment by progfolio** [comment]
> @progfolio Will this be the recommended way of upgrading built-in packages going
> forward?

For now, but it will change.
Note this does not apply to all built-in packages.
Unloading before activation should not be necessary for packages which are not pre-loaded.

> While I understand that Elpaca is now doing the _correct_ thing w.r.t.
> versioned dependencies, it feels like this warrants a more
> ergonomic/comprehensive solution. Do you see Elpaca eventually abstracting over
> the build steps [proposed
> above](https://github.com/progfolio/elpaca/issues/216#issuecomment-1868444883) -
> e.g. by exposing a new keyword?

I'd like to have *fewer* keywords in the end, to be honest.
For example, I'd like to remove `:pre-build` and `:post-build`, as they're just special cases of `:build`.

I have a WIP progress branch for allowing the `:build` keyword to accept a function which will eventually replace the limited `(:not...)` DSL it accepts now. This will allow one to use elisp to determine the build steps. e.g.

Syntactic sugar:

``` emacs-lisp
;;NOT CURRENTLY IMPLEMNETED
(elpaca (seq :build ((append (butlast (elpaca-build-steps e) '(+elpaca-unload-seq 'elpaca-activate-package))))))
```

Which is equivalent to:

```emacs-lisp
;;NOT CURRENTLY IMPLEMNETED
(elpaca (seq :build (lambda (e)
                       (append (butlast (elpaca-build-steps e) '(+elpaca-unload-seq elpaca-activate-package))))))
```

The unload function could be made more generic as well.

> It doesn't feel quite right to have to reference a bunch of internal Elpaca functions to achieve this.

The build steps will be eventually be renamed to be public symbols.
As will other parts of the API which prove useful in authoring recipes.
It's better to start with them being private and see which are needed, though.

> Also, I am not currently able to apply the solution proposed for seq to other
> built-in packages such as Eldoc (recent versions of Eglot require a higher
> version of Eldoc than is provided in Emacs 29.1). Elpaca clones the entirety of
> https://github.com/emacs-mirror/emacs into the local `elpaca/repos/eldoc`
> directory but something happens with the autoloads so that dependent code can't
> locate any of the Eldoc functions.

Please open a separate issue with the details and I'll look into it.
 
> Is there a way to revert to the old behaviour and ignore version requirements
> for a particular package?

Yes. You can skip the check-version build step.
Caveat emptor: This leaves the package vulnerable to hard-to-debug errors which crop up when version dependencies are not met.
See the following test case:

<details open><summary>Test Case</summary>

[How to run this test?](https://github.com/progfolio/elpaca/wiki/Troubleshooting#the-elpaca-test-macro)

```emacs-lisp
(elpaca-test
  :init (elpaca (pass :build (:not elpaca--check-version)))
  (elpaca-wait)
  (elpaca-test-log "#unique"))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>61c8b9b HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.6</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0) of 2023-12-18</td>
<tr><td>git</td><td>git version 2.43.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
Elpaca menu item cache discarded due to version change.
  INFO     Scraping files for loaddefs... 
  INFO     Scraping files for loaddefs...done
  GEN      ../elpaca-autoloads.el
Cloning into '/tmp/elpaca.Fs8b7N/elpaca/repos/elpaca'...
Your branch is up to date with 'origin/master'.
Checking /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-info.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-log.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-manager.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-menu-elpa.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-menu-melpa.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-menu-org.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-process.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-test.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca-ui.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/elpaca.el...
Checking /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/doc...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/doc/early-init.el...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/doc/init.el...
Checking /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/extensions...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/extensions/elpaca-use-package.el...
Checking /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/images...
Checking /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/test...
Compiling /tmp/elpaca.Fs8b7N/elpaca/repos/elpaca/test/elpaca-test.el...
Done (Total of 11 files compiled, 3 skipped in 4 directories)

Downloading MELPA recipes...
Downloading MELPA recipes...100%
Downloading NonGNU-devel ELPA...
Downloading GNU-devel ELPA...
Downloading NonGNU ELPA...
Downloading GNU ELPA...
elpaca-log: "#unique"
"
elpaca                         finished             ✓ 0.599 secs                                                                     06.479340
s                              finished             ✓ 0.732 secs                                                                     07.260736
dash                           finished             ✓ 1.433 secs                                                                     07.963785
f                              finished             ✓ 1.966 secs                                                                     08.169420
compat                         finished             ✓ 1.267 secs                                                                     08.181027
with-editor                    finished             ✓ 1.818 secs                                                                     08.394215
password-store                 finished             ✓ 2.449 secs                                                                     08.655798
password-store-otp             finished             ✓ 2.682 secs                                                                     08.887417
pass                           finished             ✓ 3.275 secs                                                                     09.156129
"

 Test Env

Elpaca 61c8b9b HEAD -> master, origin/master, origin/HEAD
installer:      0.6
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0)
 of 2023-12-18
git --version:  git version 2.43.0
```
</details>

>As an example of where this would be useful, the
> `pass` package [requires version 0.1.5 of the password-store-otp
> package](https://github.com/NicolasPetton/pass/blob/ed7031c5c33a384d07da2d15c9d5f854027a26a2/pass.el#L9)
> but AFAICT no such version exists (latest version is 0.1.0 and the code hasn't
> been updated for 3 years).

Please see the comment [here](https://github.com/progfolio/elpaca/issues/229#issuecomment-1876158980) for password-store-otp specifically.

> seemingly unnecessary version constraint but with the recent changes
> to Elpaca I'm unable to install it.

That's because the package incorrectly advertises 0.1.0 in its Version header field in the main elisp file, but has since made releases solely via git tagging.
It should be fixed upstream, but Elpaca is already able to workaround the lack of maintenance. See the comment linked above.

Please bear in mind Elpaca is still version 0.0.0, so there will be breaking changes with the aim of refining the design. However, you can `:pin` Elpaca itself, or provide the `:ref` you're interested in, etc. The current state should be flexible enough to ease the pain of development and allow one to engage with it when they have time.

Does that help?

**Comment by ashlineldridge** [comment]
Thanks for the explanation. Yes, that helps and understand there will be breaking changes. `:pin`-ing Elpaca itself is a good suggestion. I will raise a separate issue for my Eldoc problems.

**Comment by progfolio** [comment]
> ;; Workaround an oversight
> (when (version= emacs-version "29.4")
>   (setf (alist-get 'transient package--builtin-versions) '(0 4 3)))

Thanks for the reminder. I've added an entry for 29.4.
This should no longer be necessary after updating Elpaca.
  
> ;; Reduce to minimum set of ignores, to upgrade all upgradables on demand
> (cl-loop for item in elpaca-ignored-dependencies
>          when (ignore-errors (elpaca-repo-dir (elpaca-recipe item)))
>          do (setq elpaca-ignored-dependencies
>                   (delq item elpaca-ignored-dependencies)))
> (add-to-list 'elpaca-ignored-dependencies 'cl-lib)
> (add-to-list 'elpaca-ignored-dependencies 'cl-generic)
> (add-to-list 'elpaca-ignored-dependencies 'nadvice)
  
A simpler approach might be:

```emacs-lisp
(defvar +elpaca-core-updates
  '(eglot eldoc erc...) ;; Whichever packages you'd like here
  "A list of items for core packages to update automatically if they are dependencies of other declared packages.")

(setq elpaca-ignored-dependencies
      (cl-set-difference elpaca-ignored-dependencies +elpaca-core-updates))
```

> ;; These CAN be upgraded, but they clone the entire GNU Emacs repo (250 MB),
> ;; so better choose your pick

Related: https://github.com/progfolio/elpaca/issues/300#issuecomment-2081701038

> Discovered a few interesting factoids.
> 
> 3. I figure Emacs core is using `let-alist`, `seq` and `map` before ever loading
>    your init.el, because these must be either ignored or upgraded to avoid that
>    warning mentioned in this thread, "⛔ Warning (emacs): seq loaded before
>    Elpaca activation".
  
  See [loadup.el](https://git.savannah.gnu.org/cgit/emacs.git/tree/lisp/loadup.el).
  Some packages are dumped as part of the Emacs image so they do not need to be loaded at startup.

> 4. If you're curious, the actual set of packages that can safely & sanely be
>    taken off `elpaca-ignored-dependencies` boils down to just 9. They are
> 
>    ```
>    '(transient use-package tramp seq project org idlwave faceup bind-key)
>    ```

I recommend approaching these on a case by case basis instead of upgrading them automatically.

**Comment by meedstrom** [comment]
Thank you for the info! 

Removed the 29.4 bit.

Yes, it's looking like most of the dependencies are good to ignore. I suspect you'll run into people overreacting again and again because of mainly eglot and transient.

Regardless, I'll buck your recommendation until it bites me in the ass :)

Simplified what I had above to this:

```elisp
(setq elpaca-ignored-dependencies
      (append
       ;; Downgrades-in-disguise
       '(cl-lib cl-generic nadvice use-package bind-key)
       ;; No repo, or repo is the GNU Emacs repo (250 MB download)
       (cl-loop
        for dep in elpaca-ignored-dependencies
        when (let ((repo (plist-get (elpaca-recipe dep) :repo)))
               (or (not repo)
                   (equal repo "https://github.com/emacs-mirror/emacs")))
        collect dep)))
```

plus the special form for upgrading `seq`.
